Update reader width live when updating in reader view

chancez opened the pull request:
Basically, instead of updating the width of the image after clicking `ok`, the width is adjusted live, which makes it easier to figure out the right value for the manga being viewed. 

Ideally we could revert the changes with the `cancel` button, but to do that requires tracking the value prior to the updates. With how the reader is designed currently, everything is using `settings` directly and so it's tricky to separate the settings being used to draw the component and the settings that are saved. So instead I've disabled the cancel button when `liveUpdate` is true, since settings are updated immediately.

While implementing this, I realized we can't update the backend every time we re-render when `liveUpdate` is true, so I've updated the logic for persisting settings to support skipping the API call to save the settings in the backend, and only perform the API it when the user clicks ok on the number setting.

schroda commented on src/screens/Reader.tsx:
```diff
@@ -235,11 +235,13 @@ export function Reader() {
     };
 
-    const setSettingValue = (key: keyof IReaderSettings, value: AllowedMetadataValueTypes) => {
+    const setSettingValue = (key: keyof IReaderSettings, value: AllowedMetadataValueTypes, persist: Boolean = true) => {
```
```suggestion
    const setSettingValue = (key: keyof IReaderSettings, value: AllowedMetadataValueTypes, persist: boolean = true) => {
```

schroda commented on src/components/settings/NumberSetting.tsx:
```diff
@@ -98,6 +98,17 @@ export const NumberSetting = ({
+        [value, handleUpdate, handleSubmit, closeDialog],
+    );
+
     useEffect(() => {
```
this causes to many state updates during the live update

tbh, don't remember why I added this (I think to update the dialog value in case there was an update to it while the dialog was open?), probably can be completely removed

schroda replied:
what about this?

chancez replied:
Tbh I don't really understand `useEffect`. So just remove it?

schroda replied:
yeah, if everything works, just remove it

useEffect gets called after every render
here it would reset the current tmp dialog value to the original value in case it has changed

schroda commented on src/components/settings/NumberSetting.tsx:
```diff
@@ -177,19 +199,21 @@ export const NumberSetting = ({
-                    <Button onClick={closeDialogWithReset} color="primary">
-                        {t('global.button.cancel')}
-                    </Button>
+                    {!liveUpdate ? (
```
the `cancel` behaviour can be kept by adding a new state (e.g. `originalValue`) which only gets updated when the change gets persisted, when canceling, the setting gets reset to the `originalValue`

chancez replied:
I tried something like that, but the problem is at least in the way settings are passed as props, the value passed in initially is updated too. 

I basically had this, but the problem was `originalValue` was being updated too. Is it because I'm not using `useState`?
```
     const closeDialog = useCallback(
        (resetValue: boolean) => {
            setIsDialogOpen(false);

            if (resetValue) {
                setDialogValue(originalValue);
            }
        },
        [originalValue],
    );
```

I'll give it another try with state to see what happens.

chancez replied:
Ok I figured this out

schroda commented:
not sure if that is fixable with the way the reader and reader settings are implemented, but it does feel kinda laggy

chancez commented on src/components/settings/NumberSetting.tsx:
```diff
@@ -36,9 +36,11 @@ type BaseProps = {
     dialogDisclaimer?: string;
     valueUnit: string;
     handleUpdate: (value: number) => void;
+    handleSubmit?: (value: number) => void;
```
Ok, I changed it slightly, only 1 optional prop, just the callback and I tried to adapt your suggestion. LMK if I did it right or not

chancez commented:
@schroda Okay I took another stab, added `cancel` back and switched `handleUpdate` to the previous behavior and added a new optional `handleLiveUpdate` callback prop to update the setting without persisting.

schroda commented on src/components/settings/NumberSetting.tsx:
```diff
@@ -39,12 +39,15 @@ type BaseProps = {
     showSlider?: never;
     disabled?: boolean;
     listItemTextSx?: SxProps<Theme>;
+    handleLiveUpdate?: (value: number) => void;
```
looking at it again, do we even need this new callback, couldn't it just be an additional parameter (e.g. `persist`) of `handleUpdate` that is only true if the ok button got clicked?

chancez replied:
That feels like a weird thing to add to the generic `NumberSetting` component. Persistence isn't really something this component should need to worry about, no?

schroda replied:
maybe `persist` was a bad name, more like `okClicked`, but nevermind me, I think having 2 callbacks is easier to use

schroda commented on src/components/settings/NumberSetting.tsx:
```diff
@@ -62,41 +65,50 @@ export const NumberSetting = ({
     );
 
-    const closeDialogWithReset = useCallback(() => closeDialog(true), [closeDialog]);
+    const cancel = useCallback(() => {
```
`cancel` and  `resetToDefault` are basically the same function, should be merged

chancez replied:
They're similar, but not the same. They call the same functions, but the values and conditions are different. Do you have any recommendations about they should be combined?

schroda replied:
you could do a function that gets passed an optional value to reset to as an argument

schroda commented on src/screens/Reader.tsx:
```diff
@@ -235,11 +235,13 @@ export function Reader() {
     };
 
-    const setSettingValue = (key: keyof IReaderSettings, value: AllowedMetadataValueTypes) => {
+    const setSettingValue = (key: keyof IReaderSettings, value: AllowedMetadataValueTypes, persist?: boolean) => {
```
```suggestion
    const setSettingValue = (key: keyof IReaderSettings, value: AllowedMetadataValueTypes, persist: boolean = true) => {
```

schroda commented on src/screens/settings/DefaultReaderSettings.tsx:
```diff
@@ -39,10 +39,12 @@ export function DefaultReaderSettings() {
-        requestUpdateServerMetadata(convertToGqlMeta(metadata)! ?? {}, [[key, value]]).catch(() =>
-            makeToast(t('reader.settings.error.label.failed_to_save_settings'), 'warning'),
-        );
+    const setSettingValue = (key: keyof IReaderSettings, value: AllowedMetadataValueTypes, persist?: boolean) => {
```
```suggestion
    const setSettingValue = (key: keyof IReaderSettings, value: AllowedMetadataValueTypes, persist: boolean = true) => {
```

schroda commented on src/components/settings/NumberSetting.tsx:
```diff
@@ -39,12 +39,15 @@ type BaseProps = {
     Required<Pick<BaseProps, 'defaultValue' | 'minValue' | 'maxValue'>> & { showSlider: true };
 
-type Props = BaseProps | PropsWithSlider;
+type LiveUpdateProps = (BaseProps | PropsWithSlider) & { handleLiveUpdate: (value: number) => void };
```
since this other prop was removed I don't think the extra type is need anymore

chancez commented on src/components/settings/NumberSetting.tsx:
```diff
@@ -62,41 +65,50 @@ export const NumberSetting = ({
-        },
-        [value, handleUpdate, closeDialog],
-    );
+    const submit = useCallback(() => {
```
So do I just use a regular callback function?

schroda replied:
you can just remove useCallback

chancez commented:
Pushed after rebasing and fixing what I could so I can get a re-review, but need to test things again.

chancez commented:
Ok should be good, I tested and it seems to be working as expected.

schroda commented on src/components/settings/NumberSetting.tsx:
```diff
@@ -62,45 +63,50 @@ export const NumberSetting = ({
+        setIsDialogOpen(false);
+    }, [originalValue, handleUpdate]);
 
-            const didValueChange = value !== newValue;
```
this should be kept to prevent unnecessary requests, currently clicking cancel sends a requests for the same value

schroda commented on src/components/settings/NumberSetting.tsx:
```diff
@@ -62,45 +63,50 @@ export const NumberSetting = ({
-        setDialogValue(value);
-    }, [value]);
+    const submit = () => {
+        const didValueChange = dialogValue !== originalValue;
```
this should be moved to `handleUpdate` since `cancel` and `resetToDefault` value currently send a request for the same value

chancez replied:
Hmm, previously this was in the `updateSetting` function and moving this into the `handleUpdate` prop means that all existing uses of `NumberSetting` will need to be updated, is that what's wanted?

schroda replied:
you're right, it should be handled in this component and not the prop, I mistook this for `updateValue`

I think it would be best to e.g. add a `persist` parameter to `updateValue` and then call `handleUpdate` instead of `handleLiveUpdate` in case `true` gets passed
with this everything else has to call `updateValue` to be able to change it

then the `didValueChange` check could be move there to prevent unnecessary requests/updates for all value updates and not just from `submit`

schroda commented on src/screens/Reader.tsx:
```diff
@@ -257,11 +257,13 @@ export function Reader() {
-        requestUpdateMangaMetadata(manga, [[key, value]]).catch(() =>
-            makeToast(t('reader.settings.error.label.failed_to_save_settings'), 'warning'),
-        );
+        if (persist === undefined || persist) {
```
```suggestion
        if (persist) {
```

schroda commented on src/screens/settings/DefaultReaderSettings.tsx:
```diff
@@ -39,10 +39,12 @@ export function DefaultReaderSettings() {
-            makeToast(t('reader.settings.error.label.failed_to_save_settings'), 'warning'),
-        );
+    const setSettingValue = (key: keyof IReaderSettings, value: AllowedMetadataValueTypes, persist: boolean = true) => {
+        if (persist === undefined || persist) {
```
```suggestion
        if (persist) {
```
